crates/ruff/src/rules/ruff/rules/unnecessary_list_allocation_for_first_element.rs

```diff
@@ -0,0 +1,198 @@
+    subscript: &Expr,
+) {
+    let Expr::Subscript(ast::ExprSubscript { value, slice, range, .. }) = subscript else {
+        return;
+    };
+
+    let (indexes_first_element, in_slice) = classify_subscript(slice);
+    if !indexes_first_element {
+        return;
+    }
+    let Some(iter_name) = get_iterable_name(checker, value) else {
+        return;
+    };
+
+    let mut diagnostic = Diagnostic::new(
+        UnnecessaryListAllocationForFirstElement::new(iter_name.to_string()),
+        *range,
+    );
+
+    if checker.patch(diagnostic.kind.rule()) {
+        let replacement = if in_slice {
+            format!("[next(iter({iter_name}))]")
+        } else {
+            format!("next(iter({iter_name}))")
+        };
+
+        diagnostic.set_fix(Fix::suggested(Edit::range_replacement(replacement, *range)));
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
+
+/// Fetch the name of the iterable from a list expression if the expression returns an unmodified list
+/// which can be sliced into.
+fn get_iterable_name<'a>(checker: &mut Checker, expr: &'a Expr) -> Option<&'a str> {
+    match expr {
+        Expr::Call(ast::ExprCall { func, args, .. }) => {
+            let Some(id) = get_name_id(func.as_ref()) else {
+                return None;
+            };
+            if !(id == "list" && checker.semantic().is_builtin("list")) {
+                return None;
+            }
+
+            let Some(Expr::Name(ast::ExprName { id: arg_name, .. })) = args.first() else {
+                return None;
+            };
+
+            Some(arg_name.as_str())
+        }
+        Expr::ListComp(ast::ExprListComp {
+            elt, generators, ..
+        }) => {
+            // If the `elt` field is anything other than a [`Expr::Name`], we can't be sure that it
+            // doesn't modify the elements of the underlying iterator - for example, `[i + 1 for i in x][0]`.
+            if !matches!(elt.as_ref(), Expr::Name(ast::ExprName { .. })) {
+                return None;
+            }
+
+            // If there's more than 1 generator, we can't safely say that it fits the diagnostic conditions -
+            // for example, `[(i, j) for i in x for j in y][0]`.
+            if generators.len() != 1 {
+                return None;
+            }
+
+            let generator = &generators[0];
+            // Ignore if there's an `if` statement in the comprehension, since it filters the list.
+            if !generator.ifs.is_empty() {
+                return None;
+            }
+
+            let Some(arg_name) = get_name_id(&generator.iter) else {
+                return None;
+            };
+            Some(arg_name)
+        }
+        _ => None,
+    }
+}
+
+/// Check that the slice [`Expr`] is functionally equivalent to slicing into the first element. The
+/// first `bool` checks that the element is in fact first, the second checks if it's a slice or an
+/// index.
+fn classify_subscript(expr: &Expr) -> (bool, bool) {
```

Comment:
Do you think we could use `enum` instead of `bool`s as otherwise it becomes a bit difficult to read the code without checking the documentation. Or, maybe even a struct with 2 fields.